### PR TITLE
[chore] work on locate microservice

### PR DESCRIPTION
**Pull request by danielogen**
- **What does this PR do?**

- [X] This PR adds multi-tenancy to the `locate` microservice.
- [X] Environment segregation 
- [X] Lots of code refactoring  

- **What JIRA task is related to this PR?**
The link to the JIRA board task is [here](https://airqoteam.atlassian.net/browse/PLAT-73).
- **How do I test out this PR?**
Git clone this repo `cd AirQo-api/src/locate`. 

[env file](https://drive.google.com/drive/u/1/folders/1zr4oDE7bWC_pCLx34dpqae2cd0IMVA-_)

**using docker**
**Production:** `docker build -t prod-locate --target=production .` and `docker run -it -p 5555:4001 prod-locate`
**Staging:** `docker build -t stage-locate --target=staging .` and `docker run -it -p 5555:4001 stage-locate`

**locally**
 Create and activate a python virtual env. Using: `python3 -m venv env`, `source env/bin/activate` and `pip install -r requirements.txt`. Now run `flask run` to test out the points using any REST client of your choice.

NB: By default, the `.env` file points to the development DB. You are advised to use docker if you wish to utilize `production` or `staging` DB. In all case checkout out `http://127.0.0.0:<port_number>/health` endpoint to confirm the DB you will be hitting. 

- **Which endpoints should or are ready for testing?**
All endpoints as shown [in this documentation](https://docs.google.com/spreadsheets/d/1-NGTDBtTcIi8Fg2UYSWjcK6geKrZdopz5kxRFsCEMPs/edit#gid=562835484) You just need to add the tenant parameter to each URL e.g. 

- [POST] http://localhost:5555/api/v1/locate/map/create?tenant=airqo
- [GET] http://localhost:5555/api/v1/locate/map/get?tenant=airqo&userId=<USER_ID>
- [GET] http://localhost:5555/api/v1/locate/map/get?tenant=airqo&userId=<USER_ID>&spaceName=<SPACE_NAME>
- [PUT] http://localhost:5555/api/v1/locate/map/update?tenant=airqo&userId=<USER_ID>&spaceName=<SPACE_NAME>
- [DELETE] http://localhost:5555/api/v1/locate/map/delete?tenant=airqo&userId=<USER_ID>&spaceName=<SPACE_NAME>

### CODE REVIEW

File: src/locate/app.py [hunk 1]
```diff
@@ -1,19 +1,39 @@
...
+    app.register_blueprint(monitor_bp)
+    
+
+    return app
+
+
+application = create_app(os.getenv("FLASK_ENV"))
+
+if __name__ == '__main__':
+    application.run(debug=True)
```
Comment (ja-odur) [inline src/locate/app.py:39]: Can the debug value be put in the `env` or generated depending on the environment.

Comment (danielogen) [reply src/locate/app.py]: True... this has been resolved

**Review by lilly-muyama — CHANGES_REQUESTED** [review]
1. Proper error message for when an organization doesn't exist for some of the endpoints. Right now, it returns `invalid polygon`.
2. User id should be considered in all endpoints that include saving/updating/deleting locate maps.

**Comment by ja-odur** [comment]
The endpoint `POST http://localhost:5555/api/v1/map/parishes?tenant=airqo`  with no data supplied in the body, The response is "No input data provided". Could we make this more informative by maybe stating the required fields.
Also need to validate the inputs to avoid internal server errors. for instance supplying the `polygon` dict with missing `coordinate` key throws a `KeyError` on the server.

**Review by Baalmart — COMMENTED** [review]
- [ ] **Recommend Parishes:** Works well for AirQo organisation irrespective of letter casing. Responds with "invalid polygon" for other tenants.

- [ ] **Save Map:** Could we make the space name unique for a specific user? Saving map using the same request body saves without any validation. 

- [ ] **Delete:** Deletes the planning space accordingly - continues to delete even for non existent (deleted) planning spaces. Tenancy properly implemented.

**Review by Baalmart — COMMENTED** [review]
@danielogen 
Just resolve a few (1 or 2) critical recommendations from [these issues](https://github.com/airqo-platform/AirQo-api/issues/173) and we then merge..... 💪

**Review by Priscilla-AL — COMMENTED** [review]
Finally able to save map - in the staging environment

**Review by Baalmart — COMMENTED** [review]
Just merged staging to this, how is this PR looking?
Does it have any dependencies in form of other PRs?

**Review by Baalmart — APPROVED** [review]
Looks good to me.

**Review by sserurich — APPROVED** [review]
looks and works good, however it's hard to know what the numbers like 5 in the  api route (request url ) mean and also the name for the planning space.

**Comment by Baalmart** [comment]
> looks and works good, however it's hard to know what the numbers like 5 in the api route (request url ) mean and also the name for the planning space.

I guess we could leave this as a feature request among issues? @sserurich Just to ensure that we do not miss it after the PR is merged in near future?

**Comment by Baalmart** [comment]
> > looks and works good, however it's hard to know what the numbers like 5 in the api route (request url ) mean and also the name for the planning space.
> 
> I guess we could leave this as a feature request among issues? @sserurich Just to ensure that we do not miss it after the PR is merged in near future?

Just noticed that we already have this[ issue 277](https://github.com/airqo-platform/AirQo-api/issues/277) in place so no need to recreate, thanks.

**Pull request merged**